core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -160,7 +202,22 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
         if (!reroute(allocation)) {
             return new RoutingAllocation.Result(false, clusterState.routingTable());
         }
-        return new RoutingAllocation.Result(true, new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        RoutingTable routingTable = new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData());
+        RoutingAllocation.Result result = new RoutingAllocation.Result(true, routingTable);
+        logClusterHealthStateChange(
+                new ClusterStateHealth(clusterState),
+                new ClusterStateHealth(clusterState.getMetaData(), routingTable),
+                reason
+        );
+        return result;
+    }
+
+    private void logClusterHealthStateChange(ClusterStateHealth previousStateHealth, ClusterStateHealth newStateHealth, String reason) {
```

Comment:
I see were you are headed and I also see your point. What is appealing to me in the current implementation is the symmetry in the signature: 

`void logClusterHealthStateChange(ClusterStateHealth previousStateHealth, ClusterStateHealth newStateHealth, String reason)`

Callers have to provide the previous and new health and a reason. This signature is straightforward to me. 

This would be the updated version:

`void logClusterHealthStateChange(ClusterStateHealth previousStateHealth, Result newResult, String reason)`

Admittedly, the method is now easier to use (no need to create a new `ClusterStateHealth`instance from outside) I think it makes the signature less obvious. Btw, I'd call the parameter `newResult` in order to communicate which result is needed.